feat: add ability to sort by last login

yemkareems opened the pull request:
<!--
  - 🚨 SECURITY INFO
  -
  - Before sending a pull request that fixes a security issue please report it via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/). This allows us to coordinate the fix and release without potentially exposing all Nextcloud servers and users in the meantime.
-->

* Resolves: # <!-- related github issue -->

## Summary


## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

yemkareems committed:
feat: add ability to sort users by last_login, uid or displayName. one of these needs to be passed as orderBy and sort can be ASC or DESC

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
feat: cs fix run on the changed files and default order by last_login desc. Also last_login sort logic changed

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: sort the user getDisplayNames based on lastLogin. Default sort order is lastLogin DESC

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: cs fix and psalm ci related changes

Signed-off-by: yemkareems <yemkareems@gmail.com>

Pytal commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -130,15 +130,15 @@ public function __construct(
 	 * 200: Users returned
 	 */
-	public function getUsers(string $search = '', ?int $limit = null, int $offset = 0): DataResponse {
+	public function getUsers(string $search = '', ?int $limit = null, int $offset = 0, string $orderBy = 'lastLogin', string $sort = 'DESC'): DataResponse {
```
```suggestion
	public function getUsers(string $search = '', ?int $limit = null, int $offset = 0, string $orderBy = 'uid', string $sort = 'ASC'): DataResponse {
```
We should not break the existing sort order

Pytal commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -167,19 +167,21 @@ public function getUsers(string $search = '', ?int $limit = null, int $offset =
 	 * @param int|null $limit Limit the amount of groups returned
 	 * @param int $offset Offset for searching for groups
+	 * @param string $orderBy Field to order the results with
+	 * @param string $sort ASC or DESC
```
```suggestion
	 * @param string $sort asc or desc
```
Would use lowercase like our other API endpoints

Pytal commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -167,19 +167,21 @@ public function getUsers(string $search = '', ?int $limit = null, int $offset =
 	 * 200: Users details returned
 	 */
-	public function getUsersDetails(string $search = '', ?int $limit = null, int $offset = 0): DataResponse {
+	public function getUsersDetails(string $search = '', ?int $limit = null, int $offset = 0, string $orderBy = 'lastLogin', string $sort = 'DESC'): DataResponse {
```
```suggestion
	public function getUsersDetails(string $search = '', ?int $limit = null, int $offset = 0, string $orderBy = 'uid', string $sort = 'asc'): DataResponse {
```
Same to not break the sort order + lowercase sort param

kesselb commented on lib/private/User/Database.php:
```diff
@@ -264,27 +264,36 @@ public function getDisplayName($uid): string {
 				$query->expr()->eq('userid', 'uid'),
-				$query->expr()->eq('appid', $query->expr()->literal('settings')),
-				$query->expr()->eq('configkey', $query->expr()->literal('email')))
+				$query->expr()->eq('appid', $query->expr()->literal($appId)),
```
This change breaks searching by email.

1) Open user management
2) Click on the search icon on the upper right next to your profile picture
3) Enter an email address of a user

=> No result on this branch

The left join is there to look up a user by the email address. 
A subselect for lastLogin (similar to the preferences join) should work.

kesselb replied:
```
Index: lib/private/User/Database.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lib/private/User/Database.php b/lib/private/User/Database.php
--- a/lib/private/User/Database.php	(revision 1b86b6c7f67d1fbc2936a76289e93df2a6e63445)
+++ b/lib/private/User/Database.php	(date 1715447859568)
@@ -275,25 +275,32 @@
 
 		$query = $this->dbConn->getQueryBuilder();
 
-		$appId = 'settings';
-		$configKey = 'email';
-		if($orderBy == 'lastLogin') {
-			$appId = 'login';
-			$configKey = 'lastLogin';
+		if ($orderBy === 'lastLogin') {
+			$lastLoginSubSelect = $this->dbConn->getQueryBuilder();
+			$lastLoginSubSelect->select('configvalue')
+				->from('preferences', 'p2')
+				->where($lastLoginSubSelect->expr()->andX(
+					$lastLoginSubSelect->expr()->eq('p2.userid', 'uid'),
+					$lastLoginSubSelect->expr()->eq('p2.appid', $lastLoginSubSelect->expr()->literal('login')),
+					$lastLoginSubSelect->expr()->eq('p2.configkey', $lastLoginSubSelect->expr()->literal('lastLogin')),
+				));
+			$orderByExpression = $query->createFunction('(' . $lastLoginSubSelect->getSQL() .')');
+		} else {
+			$orderByExpression = $query->func()->lower('displayname');
 		}
 
-		$query->select('uid', 'displayname')
+		$query->select('uid', 'displayname', $orderByExpression)
 			->from($this->table, 'u')
 			->leftJoin('u', 'preferences', 'p', $query->expr()->andX(
 				$query->expr()->eq('userid', 'uid'),
-				$query->expr()->eq('appid', $query->expr()->literal($appId)),
-				$query->expr()->eq('configkey', $query->expr()->literal($configKey)))
+				$query->expr()->eq('appid', $query->expr()->literal('settings')),
+				$query->expr()->eq('configkey', $query->expr()->literal('email')))
 			)
 			// sqlite doesn't like re-using a single named parameter here
 			->where($query->expr()->iLike('uid', $query->createPositionalParameter('%' . $this->dbConn->escapeLikeParameter($search) . '%')))
 			->orWhere($query->expr()->iLike('displayname', $query->createPositionalParameter('%' . $this->dbConn->escapeLikeParameter($search) . '%')))
 			->orWhere($query->expr()->iLike('configvalue', $query->createPositionalParameter('%' . $this->dbConn->escapeLikeParameter($search) . '%')))
-			->orderBy($query->func()->lower('configvalue'), $sort)
+			->orderBy($orderByExpression, $sort)
 			->addOrderBy('uid_lower', 'ASC')
 			->setMaxResults($limit)
 			->setFirstResult($offset);

```

kesselb replied:
If getDisplayNames is called from getUsers (the route the user management xhr request is using) the list is sorted again in: https://github.com/nextcloud/server/blob/4560ddd706d2ced7fb47487cf57095513ae2de70/lib/private/User/Database.php#L428

yemkareems replied:
Changed the query to support  email based search and removed the sort in Database.php

yemkareems replied:
@kesselb Reverted these changes and implemented a new users/recent endpoint to get the users sorted based on lastLogin asc or desc

kesselb commented on lib/private/User/Manager.php:
```diff
@@ -288,20 +288,49 @@ public function checkPasswordNoLogging($loginName, $password) {
 				}
 			}
 		}
+		switch ($orderBy.' '.$sort) {
```
Sorting here only works with $limit and $offset null, for paginated requests the result will be wrong.

kesselb commented on lib/private/User/Manager.php:
```diff
@@ -324,9 +353,40 @@ public function searchDisplayName($pattern, $limit = null, $offset = null) {
-		usort($users, function (IUser $a, IUser $b) {
-			return strcasecmp($a->getDisplayName(), $b->getDisplayName());
-		});
+		switch ($orderBy.' '.$sort) {
```
Sorting here only works with $limit and $offset null, for paginated requests the result will be wrong.

yemkareems committed:
fix: change orderBy and sort to sortMode and sortOrder. default it to uid asc. enable email search by changing query

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: rebased the branch with master and resolved conflicts
fix: added a new endpoint users/recent and getting users based on last login info in the same. Reverted old code that was breaking LDAP

Signed-off-by: yemkareems <yemkareems@gmail.com>

Pytal commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -300,6 +300,96 @@ public function getDisabledUsersDetails(?int $limit = null, int $offset = 0): Da
+	 *
+	 * 200: Users details returned based on last logged in information
+	 */
+	public function getLastLoggedInUsers(string $search = '',
+										 ?int   $limit = null,
+										 int    $offset = 0,
+										 string $sortMode = 'lastLogin',
+										 string $sortOrder = 'desc'
```
```suggestion
	public function getLastLoggedInUsers(string $search = '',
										 ?int   $limit = null,
										 int    $offset = 0,
										 string $sortOrder = 'desc'
```
Let's keep it simple as we only need sorting by last login

yemkareems replied:
users/recent will work as it is without any params (default sort is by lastLogin desc) and in case asc/desc is required frontend can pass users/recent?offset=0&limit=25&search=&sortOrder=asc&sortBy=lastLogin params accordingly.

yemkareems committed:
fix: removed csrf check for the recent end point and ran cs fix

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: search and searchDisplayNames reverted to how it was initially as sort and order are not required here

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: search and searchDisplayNames reverted to how it was initially as sort and order are not required here

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems commented on lib/public/IUserManager.php:
```diff
@@ -126,7 +126,7 @@ public function checkPassword($loginName, $password);
 	 * @since 8.0.0
 	 */
-	public function search($pattern, $limit = null, $offset = null);
+	public function search($pattern, $limit = null, $offset = null, $orderBy = 'uid', $sort = 'ASC');
```
@kesselb reverted these changes and implemented a new endpoint that will not break the LDAP

yemkareems committed:
fix: doc blocks added

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: since added to doc blocks

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: removed the params related to sortMode and order since it sorts by lastLogin

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: removed NoCSRFRequired used for testing locally

Signed-off-by: yemkareems <yemkareems@gmail.com>

come-nc commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -265,6 +265,91 @@ public function getDisabledUsersDetails(string $search = '', ?int $limit = null,
+						fn (IUser $user): string => $user->getUID(),
+						array_filter(
+							$group->searchUsers($search, ($tempLimit === null ? null : $tempLimit - count($users))),
+							fn (IUser $user): bool => !$user->isEnabled()
```
Why is this filtering disabled users?

yemkareems replied:
@come-nc removed the references to old disabled users code

come-nc commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -265,6 +265,91 @@ public function getDisabledUsersDetails(string $search = '', ?int $limit = null,
+		if ($this->groupManager->isAdmin($uid)) {
+			$users = $this->userManager->getUsersSortedByLastLogin($limit, $offset, $search);
+			$users = array_map(fn (IUser $user): string => $user->getUID(), $users);
+		} elseif ($subAdminManager->isSubAdmin($currentUser)) {
+			$subAdminOfGroups = $subAdminManager->getSubAdminsGroups($currentUser);
+
+			$users = [];
+			/* We have to handle offset ourselve for correctness */
+			$tempLimit = ($limit === null ? null : $limit + $offset);
+			foreach ($subAdminOfGroups as $group) {
+				$users = array_merge(
+					$users,
+					array_map(
+						fn (IUser $user): string => $user->getUID(),
+						array_filter(
+							$group->searchUsers($search, ($tempLimit === null ? null : $tempLimit - count($users))),
+							fn (IUser $user): bool => !$user->isEnabled()
+						)
+					)
+				);
+				if (($tempLimit !== null) && (count($users) >= $tempLimit)) {
+					break;
+				}
+			}
+			$users = array_slice($users, $offset);
+		}
```
This whole branch is returning the wrong thing.

come-nc replied:
Maybe we should not support subadmins here, because perf of sorting all group members by last login may be terrible? @artonge ?

yemkareems replied:
@come-nc I had a call with @artonge on this and it was decided we will support last login sorting for admin alone and not for subadmin and groups

yemkareems replied:
Sub admin and groups are not supported. Hence removed this

come-nc commented on lib/private/AllConfig.php:
```diff
@@ -491,6 +491,51 @@ public function getUsersForUserValue($appName, $key, $value) {
+
+		$query = $this->connection->getQueryBuilder();
+
+		$lastLoginSubSelect = $this->connection->getQueryBuilder();
+		$lastLoginSubSelect->select('configvalue')
+			->from('preferences', 'p2')
+			->where($lastLoginSubSelect->expr()->andX(
+				$lastLoginSubSelect->expr()->eq('p2.userid', 'uid'),
+				$lastLoginSubSelect->expr()->eq('p2.appid', $lastLoginSubSelect->expr()->literal('login')),
+				$lastLoginSubSelect->expr()->eq('p2.configkey', $lastLoginSubSelect->expr()->literal('lastLogin')),
+			));
+		$orderByExpression = $query->createFunction('(' . $lastLoginSubSelect->getSQL() .')');
```
This feels unefficient to me, but maybe I’m wrong, wouldn’t a second join be faster ?
ping @Altahrim

yemkareems replied:
@come-nc, @Altahrim Intially this was written in getDisplayNames and that broke the search by email. I ll refactor this query only to sort by lastLogin configvalue desc

yemkareems replied:
Refactored this query to similar to getDisplayNames query to make it better

come-nc commented on lib/private/User/Manager.php:
```diff
@@ -342,6 +342,36 @@ public function getDisabledUsers(?int $limit = null, int $offset = 0, string $se
+			)
+		);
+
+		$tempLimit = ($limit === null ? null : $limit + $offset);
+		foreach ($this->backends as $backend) {
+			if (($tempLimit !== null) && (count($users) >= $tempLimit)) {
+				break;
+			}
+			if ($backend instanceof IProvideEnabledStateBackend) {
+				$backendUsers = $backend->getDisabledUserList(($tempLimit === null ? null : $tempLimit - count($users)));
+				foreach ($backendUsers as $uid) {
+					$users[$uid] = new LazyUser($uid, $this, null, $backend);
+				}
+			}
+		}
```
This part has nothing to do here, it’s about disabled users.
@yemkareems please test your PR more in depth before asking for reviews.

yemkareems replied:
@come-nc Sorry about the old code leftover here. Removed the old disabled users code and added the limit and offset to the query and cleaned it up here

come-nc commented on lib/public/IConfig.php:
```diff
@@ -249,4 +249,13 @@ public function deleteAppFromAllUsers($appName);
+	 * @return array of user IDs
+	 * @since 30.0.0
+	 */
+	public function getLastLoggedInUsers($search);
```
It should also contain the limit, and maybe the offset, to be efficient.

yemkareems replied:
Added offset and limit to the query

come-nc commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -265,6 +265,91 @@ public function getDisabledUsersDetails(string $search = '', ?int $limit = null,
+	 * 200: Users details returned based on last logged in information
+	 */
+	public function getLastLoggedInUsers(string $search = '',
+		?int   $limit = null,
```
limit should not default to null, and maybe not even be nullable, otherwise you just get a list of all users.

yemkareems replied:
Will default the limit to 25

yemkareems replied:
Added offset and limit to the query

yemkareems replied:
Limit is defaulted to null and if no limit is given all users are fetched ordered by last login desc and if limit, offset and search are given the result set is fetched accordingly @come-nc @Pytal

yemkareems committed:
fix: removed references to old disabled users code. refactored query as per getDisplayNames function. limit and offset added to query. default limit set to 25.

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: removed default limit of 25. if null is given all users are fetched or if limit is given limit number of users are fetched

Signed-off-by: yemkareems <yemkareems@gmail.com>

come-nc commented on apps/provisioning_api/lib/Controller/UsersController.php:
```diff
@@ -265,6 +265,68 @@ public function getDisabledUsersDetails(string $search = '', ?int $limit = null,
 	}
 
+	/**
+	 * @NoAdminRequired
```
```suggestion
```
Since you removed subadmin handling, this should be for admins only now.

yemkareems replied:
Removed the doc block and if condition

yemkareems committed:
fix: limit and fixLimit removed. negative limit handled in controller. removed getUsersSortedByLastLogin from Manager and instead used the config in controller

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems commented on lib/private/AllConfig.php:
```diff
@@ -491,6 +491,46 @@ public function getUsersForUserValue($appName, $key, $value) {
 	}
 
+	/**
+	 * Gets the list of users based on their lastLogin info asc or desc
```
Changed comment

yemkareems commented on lib/private/AllConfig.php:
```diff
@@ -534,4 +574,12 @@ public function getUsersForUserValueCaseInsensitive($appName, $key, $value) {
+		}
+
+		return null;
+	}
```
negative limit and offset throwing up is handled in controller. So removed the same here

yemkareems commented on lib/private/User/Manager.php:
```diff
@@ -342,6 +342,21 @@ public function getDisabledUsers(?int $limit = null, int $offset = 0, string $se
+		);
+	}
+
+
```
Removed this method and used the config directly in controller

come-nc commented on lib/private/AllConfig.php:
```diff
@@ -491,6 +491,45 @@ public function getUsersForUserValue($appName, $key, $value) {
+			// sqlite doesn't like re-using a single named parameter here
+			->where($query->expr()->iLike('uid', $query->createPositionalParameter('%' . $this->connection->escapeLikeParameter($search) . '%')))
+			->orWhere($query->expr()->iLike('displayname', $query->createPositionalParameter('%' . $this->connection->escapeLikeParameter($search) . '%')))
+			->orWhere($query->expr()->iLike('configvalue', $query->createPositionalParameter('%' . $this->connection->escapeLikeParameter($search) . '%')))
```
This only makes sense to search in email, but this will not work because you join on configkey lastLogin.
So either you have to join twice with `preferences` if you want search to work in email address, or you should remove this line and search will not search in email address.

yemkareems replied:
Will add a extra join to have the search with email intact

come-nc commented on lib/private/AllConfig.php:
```diff
@@ -491,6 +491,45 @@ public function getUsersForUserValue($appName, $key, $value) {
+			->setMaxResults($limit);
+
+		$result = $query->executeQuery();
+		$displayNames = [];
```
Why is this var called $displayNames? It contains uids.
```suggestion
		$uids = [];
```

yemkareems replied:
changed it to uids

come-nc commented on lib/private/AllConfig.php:
```diff
@@ -491,6 +491,45 @@ public function getUsersForUserValue($appName, $key, $value) {
+
+		$result = $query->executeQuery();
+		$displayNames = [];
+		while ($row = $result->fetch()) {
```
```suggestion
		while (($row = $result->fetch()) !== false) {
```
You have to check against false with a strong check to not stop on a uid being `'0'` which is valid.

come-nc commented on lib/private/AllConfig.php:
```diff
@@ -491,6 +491,45 @@ public function getUsersForUserValue($appName, $key, $value) {
+
+		$query = $this->connection->getQueryBuilder();
+
+		$query->select('uid', 'displayname')
```
```suggestion
		$query->selectDistinct('uid')
```
You do not use displayname so no need to select it.
And if you use selectDistinct you can avoid using uid as key in the foreach loop below and can return a list which is better than an associative array.
You can even return $result->fetchAll in column mode.

yemkareems replied:
used selectDistinct and fetchAll

come-nc commented on lib/private/AllConfig.php:
```diff
@@ -491,6 +491,45 @@ public function getUsersForUserValue($appName, $key, $value) {
+			->setFirstResult($offset)
+			->setMaxResults($limit);
+
+		$result = $query->executeQuery();
+		$displayNames = [];
```
```suggestion
		$result = $query->executeQuery();
		$uids = $result->fetchAll(\PDO::FETCH_COLUMN);
		$result->closeCursor();
		return $uids;
```
Like that, if you only select uid with distinct you can use this simple code block instead of a while.

come-nc commented on lib/public/IConfig.php:
```diff
@@ -249,4 +249,15 @@ public function deleteAppFromAllUsers($appName);
+	 * @param int|null $limit how many records to fetch
+	 * @param int $offset from which offset to fetch
+	 * @param string $search search users based on search params
+	 * @return array of user IDs
```
```suggestion
	 * @return list<string> list of user IDs
```

yemkareems committed:
fix: query refactored to support both the search by email and sort by login

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: second join and where conditions added only when search param is available

Signed-off-by: yemkareems <yemkareems@gmail.com>

come-nc commented on lib/private/AllConfig.php:
```diff
@@ -491,6 +491,50 @@ public function getUsersForUserValue($appName, $key, $value) {
+	 * @param int|null $limit how many users to fetch
+	 * @param int $offset from which offset to fetch
+	 * @param string $search search users based on search params
+	 * @return array of user IDs
```
```suggestion
	 * @return list<string> list of user IDs
```
To please psalm

yemkareems committed:
fix: return type doc block added as per psalm

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: return type doc block added as per psalm

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
fix: return type doc block added as per psalm

Signed-off-by: yemkareems <yemkareems@gmail.com>

artonge commented:
Rebased

Pytal commented on lib/public/IConfig.php:
```diff
@@ -249,4 +249,15 @@ public function deleteAppFromAllUsers($appName);
 	public function getUsersForUserValue($appName, $key, $value);
+
+	/**
+	 * Gets the list of users based on their lastLogin info asc or desc
```
```suggestion
	 * Gets the list of users sorted by lastLogin, from most recent to least recent
```
Since we can't choose asc or desc, copied from the other method

yemkareems committed:
fix: doc block corrected

Signed-off-by: yemkareems <yemkareems@gmail.com>

artonge commented:
You probably need to fix the openapi check

kesselb reviewed (dismissed):
Why IConfig/AllConfig? :confused:

yemkareems committed:
fix: ran bash build/openapi-checker.sh and commit the result

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems commented:
> Why IConfig/AllConfig? 😕

It is mostly in the lines of the getDisabledUsers which calls IUserManager/Manager which internally calls IConfig/AllConfig. The Manager was not required because we just wanted the list of user and the LazyUser in the Manager was not required. So i did IConfig/AllConfig

kesselb commented:
`IConfig` is the public api to create/read/update/delete configuration values, either from configuration file, app config, or user preferences. 

I assume `getDisabledUsers` uses `getUsersForUserValue` because that's the way to retrieve user preferences. 

`getLastLoggedInUsers` does not belong in IConfig and should go somewhere else. The UserManager looks like a good place, but please check with Louis, and we are already using the QueryBuilder there, regrettably not via DI.

come-nc commented:
> `IConfig` is the public api to create/read/update/delete configuration values, either from configuration file, app config, or user preferences.
> 
> I assume `getDisabledUsers` uses `getUsersForUserValue` because that's the way to retrieve user preferences.
> 
> `getLastLoggedInUsers` does not belong in IConfig and should go somewhere else. The UserManager looks like a good place, but please check with Louis, and we are already using the QueryBuilder there, regrettably not via DI.

I disagree, it’s reading in the configuration table so it’s better situated in IConfig than IUserManager, no?

artonge commented:
I agree with @kesselb, this seems way too specific to be located in `IConfig`.

yemkareems commented:
> I agree with @kesselb, this seems way too specific to be located in `IConfig`.

Sure @kesselb , @artonge i ll move this to UserManager/Manager as suggested

yemkareems committed:
fix: getLastLoggedInUsers moved from AllConfig/IConfig to IUserManager/Manager

Signed-off-by: yemkareems <yemkareems@gmail.com>

provokateurin committed:
fix(OpenAPI): Regenerate

Signed-off-by: provokateurin <kate@provokateurin.de>

nickvergessen commented on lib/public/IUserManager.php:
```diff
@@ -210,4 +210,15 @@ public function getByEmail($email);
+	 * @return list<string> list of user IDs
+	 * @since 30.0.0
+	 */
+	public function getLastLoggedInUsers(?int $limit = null, int $offset = 0, string $search = ''): array;
```
Please document in https://docs.nextcloud.com/server/latest/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_30.html#id1
